### PR TITLE
IGNITE-15356 .NET: Add basic thin client

### PR DESCRIPTION
**Pull request by ptupitsyn**
* Set up the project structure
* Establish code style and enable code analysis
* Implement serialization based on `MessagePack-CSharp` and `ArrayPool.Shared`
* Implement async socket IO
* Implement basic thin client (tuple, get table, upsert, get)

https://cwiki.apache.org/confluence/display/IGNITE/IEP-78+.NET+Thin+Client

### CODE REVIEW

File: modules/platforms/dotnet/Apache.Ignite/Internal/Buffers/PooledArrayBufferWriter.cs [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+        private const int ReservedPrefixSize = 4 + 4 + 9; // Size (4 bytes) + OpCode (4 bytes) + RequestId (9 bytes)/
+
+        /** Underlying pooled array. */
+        private byte[] _buffer;
+
+        /** Index within the array. */
+        private int _index;
+
+        /** Index within the array. */
+        private int _index2;
```
Comment (isapego) [inline modules/platforms/dotnet/Apache.Ignite/Internal/Buffers/PooledArrayBufferWriter.cs:53]: Probably needs a better name.

File: modules/platforms/dotnet/Apache.Ignite/Internal/Proto/MessagePackReaderExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+            byte h = jBytes[11];
+            byte i = jBytes[10];
+            byte j = jBytes[9];
+            byte k = jBytes[8];
+
+            int a = BinaryPrimitives.ReadInt32BigEndian(jBytes[4..]);
+            short b = BinaryPrimitives.ReadInt16BigEndian(jBytes[2..]);
+            short c = BinaryPrimitives.ReadInt16BigEndian(jBytes);
+
+            return new Guid(a, b, c, d, e, f, g, h, i, j, k);
```
Comment (isapego) [inline modules/platforms/dotnet/Apache.Ignite/Internal/Proto/MessagePackReaderExtensions.cs:114]: Why GUID serialization format is that weird?

**Review by isapego — APPROVED** [review]
Overall looks good.

File: modules/platforms/dotnet/Apache.Ignite.Tests/Buffers/PooledArrayBufferWriterTests.cs [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+            var writer = bufferWriter.GetMessageWriter();
+            writer.Write(1);
+            writer.Write("A");
+            writer.Flush();
+
+            var prefixWriter = bufferWriter.GetPrefixWriter(3);
+            prefixWriter.Write(7);
+            prefixWriter.Write(8);
+            prefixWriter.Write(9);
+            writer.Flush();
```
[the remark this answers is not in the document]
Comment (ptupitsyn) [reply modules/platforms/dotnet/Apache.Ignite.Tests/Buffers/PooledArrayBufferWriterTests.cs:65]: Good catch, fixed. Flush is not always necessary, we got lucky here, but it should always be called.

File: modules/platforms/dotnet/Apache.Ignite/Internal/Buffers/PooledArrayBufferWriter.cs (hunk 1 above)
Comment (ptupitsyn) [reply modules/platforms/dotnet/Apache.Ignite/Internal/Buffers/PooledArrayBufferWriter.cs:53]: Agree, fixed.

File: modules/platforms/dotnet/Apache.Ignite/Internal/Proto/MessagePackReaderExtensions.cs (hunk 2 above)
Comment (ptupitsyn) [reply modules/platforms/dotnet/Apache.Ignite/Internal/Proto/MessagePackReaderExtensions.cs:117]: `Guid` and `UUID` use different encoding, see https://en.wikipedia.org/wiki/Universally_unique_identifier#Encoding.
XMLDoc comments updated.

**Pull request merged**